firebase-functions/src/main/java/com/google/firebase/functions/FirebaseFunctions.kt

```diff
@@ -0,0 +1,383 @@
+import java.net.URL
+import java.util.concurrent.Executor
+import javax.inject.Named
+
+/** FirebaseFunctions lets you call Cloud Functions for Firebase.  */
+class FirebaseFunctions @AssistedInject internal constructor(
+        context: Context,
+        @Named("projectId") projectId: String?,
+        @Assisted regionOrCustomDomain: String?,
+        contextProvider: ContextProvider?,
+        @param:Lightweight private val executor: Executor,
+        @UiThread uiExecutor: Executor) {
+  // The network client to use for HTTPS requests.
+  private val client: OkHttpClient
+
+  // A serializer to encode/decode parameters and return values.
+  private val serializer: Serializer
+
+  // A provider of client metadata to include with calls.
+  private val contextProvider: ContextProvider
+
+  // The projectId to use for all functions references.
+  private val projectId: String
+
+  // The region to use for all function references.
+  private var region: String? = null
+
+  // A custom domain for the http trigger, such as "https://mydomain.com"
+  private var customDomain: String? = null
+
+  // The format to use for constructing urls from region, projectId, and name.
+  private var urlFormat = "https://%1\$s-%2\$s.cloudfunctions.net/%3\$s"
+
+  // Emulator settings
+  private var emulatorSettings: EmulatedServiceSettings? = null
+
+  init {
+    client = OkHttpClient()
+    serializer = Serializer()
+    this.contextProvider = Preconditions.checkNotNull(contextProvider)
+    this.projectId = Preconditions.checkNotNull(projectId)
+    val isRegion: Boolean
+    isRegion = try {
+      URL(regionOrCustomDomain)
+      false
+    } catch (malformedURLException: MalformedURLException) {
+      true
+    }
+    if (isRegion) {
+      region = regionOrCustomDomain
+      customDomain = null
+    } else {
+      region = "us-central1"
+      customDomain = regionOrCustomDomain
+    }
+    maybeInstallProviders(context, uiExecutor)
+  }
+
+  /** Returns a reference to the callable HTTPS trigger with the given name.  */
+  fun getHttpsCallable(name: String): HttpsCallableReference {
+    return HttpsCallableReference(this, name, HttpsCallOptions())
+  }
+
+  /** Returns a reference to the callable HTTPS trigger with the provided URL.  */
+  fun getHttpsCallableFromUrl(url: URL): HttpsCallableReference {
+    return HttpsCallableReference(this, url, HttpsCallOptions())
+  }
+
+  /** Returns a reference to the callable HTTPS trigger with the given name and call options.  */
+  fun getHttpsCallable(
+          name: String, options: HttpsCallableOptions): HttpsCallableReference {
+    return HttpsCallableReference(this, name, HttpsCallOptions(options))
+  }
+
+  /** Returns a reference to the callable HTTPS trigger with the provided URL and call options.  */
+  fun getHttpsCallableFromUrl(
+          url: URL, options: HttpsCallableOptions): HttpsCallableReference {
+    return HttpsCallableReference(this, url, HttpsCallOptions(options))
+  }
```

Comment:
We could reduce this down to 2 functions by using default arguments and `@JvmOverloads`:

```kotlin
  @JvmOverloads
  fun getHttpsCallable(name: String, options: HttpsCallableOptions? = null): HttpsCallableReference {
    return HttpsCallableReference(this, name, HttpsCallOptions(options))
  }

  @JvmOverloads
  fun getHttpsCallableFromUrl(
          url: URL, options: HttpsCallableOptions? = null): HttpsCallableReference {
    return HttpsCallableReference(this, url, HttpsCallOptions(options))
  }
```